### DISCUSSION TITLE
How to prevent schema from ending in client bundles in api-routes-apollo-server-and-client

### DISCUSSION THREAD

**Discussion by fromi** [vercel/next.js#45011] [Help]
Next.js provides an example to combine apollo server & apollo client here: https://github.com/vercel/next.js/tree/canary/examples/api-routes-apollo-server-and-client

In this exemple, we have this import chain:
_app.tsx => `{ useApollo } from '../apollo/client'` => `{ schema } from '../apollo/schema'`
It means that every page will be bundled with the apollo schema... However **the schema is only used on server side** (`if (typeof window === 'undefined')` in client.tsx)

The schema basically contains all the server side code with all the renderers, so the example seems incomplete.

Based on blog posts, I worked around the issue using inline require:

```
const {SchemaLink} = require('@apollo/client/link/schema')
const schema = require('./schema').default
```

Everything was working fine... until NextJS v12. Now, require returns a Promise (https://github.com/vercel/next.js/issues/42931)
It is still working using `experimental: {esmExternals: false}`, but it feels like a mess.

I would love to see the api-routes-apollo-server-and-client example improved with a solution to this huge issue.

**Comment by icyJoseph** [comment]
Hi!

You could tell Webpack to not bundle up apollo/schema. Something like:

```js
  webpack(config, { nextRuntime }) {
    if (typeof nextRuntime === "undefined") {
      // manually ignore some local modules and node_modules
      const { IgnorePlugin } = require("webpack");
      const ignoreUserModule = new IgnorePlugin({
        resourceRegExp: /apollo\/schema/,
      });

      config.plugins.push(ignoreUserModule);
    }

    return config;
  },
```

That's make sure that during bundling Webpack doesn't follow apollo/schema on the client. `nextRuntime` is undefined for client side bundles.


Before:
 
![Screenshot 2023-01-18 at 18 08 19](https://user-images.githubusercontent.com/21013447/213248072-aaff25bf-dfa0-47fc-84ef-c97c0473f6fb.png)


After:

![Screenshot 2023-01-18 at 18 08 32](https://user-images.githubusercontent.com/21013447/213248035-a4e90a9c-2c23-4d85-8f7a-0f9b7c2d4404.png)

**Reply by fromi** [reply]
Hi, I tried that, but it does not prevent the imports from inside the schema to be bundled.
I think I would have to cascade-ignore all the dependencies starting with the schema... Is that correct?

**Reply by fromi** [reply]
I found a working solution using NormalModuleReplacementPlugin:
```
webpack: (config, {isServer}) => {
    if (!isServer) {
      config.plugins.push(new NormalModuleReplacementPlugin(/\.\/schema$/, resource => {
        resource.request = resource.request.replace(/\.\/schema/, '\.\/fake')
      }))
  }
}
```

**Reply by icyJoseph** [reply]
I guess so, do check `nextRuntime` instead of `isServer` though, but the idea is to tell webpack, to not follow through with the given path.

**Reply by fromi** [reply]
Any reason why I should favor nextRuntime?
Next.js documentation specifically states that "isServer" is made for my use case:

> The webpack function is executed twice, once for the server and once for the client. This allows you to distinguish between client and server configuration using the isServer property.

(in https://nextjs.org/docs/api-reference/next.config.js/custom-webpack-config)

**Reply by icyJoseph** [reply]
Well for your use case perhaps not, but the problem is that `isServer` is true for `middleware` as well, and in general anything that starts to use the "edge" runtime, IIRC, and it might very well be the case that your module has Node API's not present in the "Edge" runtime, so that gives even more control over where the code is bundled. 

> The webpack function is executed twice, once for the server and once for the client. This allows you to distinguish between client and server configuration using the isServer property.

The docs are great and all, but for example this little bit is not correct.

![Screenshot 2023-01-20 at 14 53 21](https://user-images.githubusercontent.com/21013447/213713482-20ae3dbe-91c8-4017-945e-835674c76129.png)

I put a `console.count` inside the webpack function, and well, it ran three times. 

It is good to question this stuff though, and it would seem like in your case, `isServer` is enough.

**Reply by fromi** [reply]
Thanks! I will keep that in mind if I have other use cases later on :)
